sips/sip-015/sip-015-network-upgrade.md

```diff
@@ -0,0 +1,1595 @@
+
+The `addrs` list contains the same PoX address values passed into the PoX smart contract:
+   * They each have type signature `(tuple (hashbytes (buff 32)) (version (buff 1)))`
+   * The `version` field can be any of the following:
+      * `0x00` means this is a pay-to-public-key-hash (p2pkh) address, and `hashbytes` is the 20-byte hash160 of a single public key
+      * `0x01` means this is a pay-to-script-hash (p2sh) address, and `hashbytes` is the 20-byte hash160 of a redeemScript script
+      * `0x02` means this is a pay-to-witness-public-key-hash-over-pay-to-script-hash (p2wpkh-p2sh) address, and `hashbytes` is the 20-byte hash160 of a p2wpkh witness script
+      * `0x03` means this is a pay-to-witness-script-hash-over-pay-to-script-hash (p2wsh-p2sh) address, and `hashbytes` is the 20-byte hash160 of a p2wsh witness script
+      * `0x04` means this is a pay-to-witness-public-key-hash (p2wpkh) address, and `hashbytes` is the 20-byte hash160 of the witness script
+      * `0x05` means this is a pay-to-witness-script-hash (p2wsh) address, and `hashbytes` is the 32-byte sha256 of the witness script
+      * `0x06` means this is a pay-to-taproot (p2tr) address, and `hashbytes` is the 32-byte sha256 of the witness script
+
+**Rationale:**  This method empowers developers to query Bitcoin state.  Stacks
+2.1 adds the ability for smart contracts to query PoX payouts directly, with applications
+for Stacking pools and other Stacking-centric programs.
+
+**Examples:**
+
+```clarity
+(get-burn-block-info? header-hash u677050) ;; Returns (some 0xe67141016c88a7f1203eca0b4312f2ed141531f59303a1c267d7d83ab6b977d8)
+(get-burn-block-info? pox-addrs u677050) ;; Returns (some (tuple (addrs ((tuple (hashbytes 0x395f3643cea07ec4eec73b4d9a973dcce56b9bf1) (version 0x00)) (tuple (hashbytes 0x7c6775e20e3e938d2d7e9d79ac310108ba501ddb) (version 0x01)))) (payout u123)))
+```
+
+### New method: `slice`
+
+* **Input Signature:** `(slice (sequence sequence_A) (left uint) (right uint))`
+* **Output Signature:** `(optional sequence_A)`
+
+The `slice` function attempts to return a sub-sequence of that starts
+at `left-position` (inclusive), and ends at `right-position`
+(non-inclusive).
+
+If `left_position`==`right_position`, the function returns an empty
+sequence.
+
+If either `left_position` or `right_position` are out of bounds OR if
+`right_position` is less than `left_position`, the function returns
+`none`.
+
+**Rationale:** This method facilitates parsing user-supplied encoded data, such
+as data from oracles and other off-chain services.  While a variant of `slice`
+could be implemented entirely in Clarity, it would be much slower and much more
+expensive than supplying a native method.
+
+**Examples:**
+
+```clarity
+(slice \"blockstack\" u5 u10) ;; Returns (some \"stack\")
+(slice (list 1 2 3 4 5) u5 u9) ;; Returns none
+(slice (list 1 2 3 4 5) u3 u4) ;; Returns (some (4))
```

Comment:
I don't think non-tech user are going to care very much about 0-indexing vs 1-indexing, since either way they'll need to learn about the concept of indexing from scratch.

We chose 0-based indexing because it's what smart contract devs already expect.

Reply:
With reference to `index-of` and `element-at`, I agree. 

Conversation can be marked as resolved.